sdk/core/azure-core/inc/azure/core/uuid.hpp

```diff
@@ -26,6 +26,7 @@ namespace Azure { namespace Core {
     static constexpr size_t UuidSize = 16;
 
     std::array<uint8_t, UuidSize> m_uuid{};
+    std::string m_uuidString = {};
```

Comment:
This change almost triples the size of a Uuid - it goes from 16 bytes to 44 bytes since a `std::string` is 28 bytes in size.

QQ: Does the performance benefit of caching the string result justify the increased storage for every Uuid? Do we know how many times `Uuid::ToString()` is called on the average instance of `Azure::Core::Uuid`?

FWIW, in a pass over the source code, the most common pattern I found was: `Uuid::CreateUuid().ToString()`.

Reply:
One other question: How do you protect `m_uuidString` from multiple threads modifying it simultaneously? The contract for all C++ runtime library types is "multiple readers, single writer", which means that if there can be multiple threads making the first call into ToString() simultaneously, you can get heap corruptions.

Reply:
> FWIW, in a pass over the source code, the most common pattern I found was: `Uuid::CreateUuid().ToString()`. 

Agreed. Given that is the most common pattern for creating a Uuid in the SDK, it doesn't make sense to optimize caching `ToString()` here.

Question: Should we change the underlying representation of Uuid from an array to a string, instead, especially since we have a single supported format? Given the SDK usage of parsing or sending Uuids as strings over the wire, it seems more useful to have the underlying representation match that.

Reply:
That is an excellent question, and I was wondering the same thing. I could totally see changing it to either a std::string or a std::array<39> (has the advantage of not having a heap allocation for a Uuid, at the cost of making ToString more expensive). 

It does make `Uuid::Create` somewhat more expensive (but since the common pattern is to immediately call `ToString` that cost is just moved around). The other negative is that it makes the `AsArray` function orders of magnitude more expensive.

Reply:
Absolutely!
Now the evaluation is what's more common, what's more of a perf impactful scenario. At the cost of making the array ctor more expensive, we make the ToString() significanly faster.

My intuition suggests making the string transformation "free" since uuids in the SDK context are typically used to parse/send data ober the wire as strings (as part of request/response creation).

I'd be curious to know your thoughts on the perf sensitivity and impact of the array case, since it is primarily used in Amqp (and less so, in all the rest of the "traditional" SDKs).

Reply:
`AsArray()` is currently only used here:
https://github.com/Azure/azure-sdk-for-cpp/blob/7a9c54d915995d9d86a2ec3aa7aa0538942fb64e/sdk/core/azure-core-amqp/src/models/amqp_value.cpp#L300-L305
https://github.com/Azure/azure-sdk-for-cpp/blob/7a9c54d915995d9d86a2ec3aa7aa0538942fb64e/sdk/core/azure-core-amqp/src/models/amqp_value.cpp#L582-L584

And we have one caller for `CreateFromArray()`, in the same file, across the entire SDK.

`Uuid::CreateUuid().ToString();` shows up in every SDK call (via the RequestIdPolicy), in AMQP/EventHubs in a couple places, keyvault samples for end customers, and public storage client APIs such as `CreateUniqueLeaseId()`.

Reply:
The array format for Uuid was added to enable existing Uuid structures to be expressed in Azure::Core::Uuid. So it's not at all surprising that the only uses of that are for AMQP.

If we change Uuid to make the array forms measurably slower then in all honesty, I'll probably change the AMQP code to use a std::array<uint8_t, 16> because I need the AMQP code to have reasonable performance (the Uuid type is used in a bunch of places in AMQP, and the wire representation of a Uuid is an array of bytes, not an array of strings).

To be completely clear: it is 100% ok if we want to optimize Azure::Core::Uuid for the string case. However if we *do* decide to do that, I would only start that work if we had evidence that it made a measurable improvement to our core scenarios. 

As I tried to make clear in my emails on the subject, I feel rather strongly that we shouldn't optimize a scenario until we have evidence that performing a particular optimization will measurably improve a particular scenario.

And similarly, I wouldn't move from Azure::Core::Uuid unless my measurements for AMQP showed that parsing and serializing the string representation for Uuid was a bottleneck.

We should always measure before optimizing.

Reply:
Sounds good. I am on the same page. This is largely exploratory/learning for me, at this point.
I am using it to drive a conversation, we should certainly measure and identify whether optimizing this leads to any performance impact to the e2e scenario of a calling a client operation.

Reply:
FYI: Using string as the underlying representation, and modifying the `CreateUuid()` ctor to be as follows, results in a **2.8x perf improvement** for `Uuid::CreateUuid().ToString();`:

```C++
// private:
  Uuid(std::string const& uuid) { m_uuidString = uuid; }

// public:
  Uuid Uuid::CreateUuid()
  {
     // ... existing implementation remains as is

    // Adding the following
    std::string s(36, '-');

    for (size_t i = 0, j = 0; j < s.size() && i < UuidSize; i++)
    {
      if (i == 4 || i == 6 || i == 8 || i == 10)
      {
        j++; // Add hyphens at the appropriate places
      }

      uint8_t highNibble = (uuid[i] >> 4) & 0x0F;
      uint8_t lowNibble = uuid[i] & 0x0F;
      s[j++] = ByteToHexChar(highNibble);
      s[j++] = ByteToHexChar(lowNibble);
    }

    return Uuid(s);
  }
```

**Before:**
Completed 276,174 operations in a weighted-average of 10s (27,615.249601 ops/s, 3.62119e-05 s/op)

**After:**
Completed 784,397 operations in a weighted-average of 10s (78,343.281357 ops/s, 1.27643e-05 s/op)

I think I can apply some of the learnings here to the existing `ToString()` function.